src/core/thread/mle.cpp

```diff
@@ -1368,6 +1388,14 @@ otError Mle::AppendAddressRegistration(Message &aMessage, AddressRegistrationMod
             continue;
         }
 
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+        // Here skips DUA which was appended already.
+        if (&addr->GetAddress() == Get<DuaManager>().GetDomainUnicastAddress())
```

Comment:
We are comparing pointers here. While this works ok (since from `GetDomainUnicastAddress()` we are returning a pointer to `mAddress` field in the same `Ip6::NetifUnicastAddress` instance as the one we would be adding to `Netif` and later get in `addr` as we iterate through the `Netif` address entries) it relies on this knowledge of internal/private  implementation/behavior of `DuaManager::GetDomainUnicastAddress()` and `Netif`. If any of these gets changed in future this code would become incorrect (and such issue would not be easily noticed).

To be safe and correct, I would suggest we do check the IP address value directly instead of checking the pointers. 

Thoughts?

Reply:
Our existing usage of `RemoveUnicastAddress` and `UnsubscribeMulticast` rely on the fact that the `Ip6::NetifUnicastAddress` instances in `NetIf` are the same members. 

For example, this code rely on the fact that only `&mAllDomainBackboneRouters` can be unsubscribed because the actual content of `mAllDomainBackboneRouters` is not correctly set in some cases.  
https://github.com/openthread/openthread/blob/7b0cb984455731bbaf2eb51188baf5d225a2ee41/src/core/backbone_router/local.cpp#L369-L375

It would be worthless if we only check IP address here but other code are not properly revised.

That being said, I would prefer pointer comparing for its efficiency. Thoughts? @librasungirl

Reply:
@simonlingoogle you raise a good question but my initial point is different (not directly related to `NetifUnicastAddress` entries being same as the one added earlier (which in my opinion is fine)).

Note that `Ip6::Address` and `Ip6::NetifUnicastAddress` are two different types:
- The `DuaManager::GetDomainUnicastAddress()` is returning an `Ip6::Address *` 
- The fact that this pointer is derived from a `Ip6::NetifUnicastAddress` instance is an internal/private implementation of `DuaManager` class.
  - Note that `mDomainUnicastAddress` is a private member of `DuaManager`.
- So using this fact from another class (`Mle` here) is not good/safe.
- If in future for any reason the `DuaManager` implementation gets changed (e.g., we save the IP address differently), it would not be easy for developer making the change to realize that that `Ip6::Address *` returned from `GetDomainUnicastAddress()` may be used  like this in other modules.

----------

Regarding `Ip6::NetifUnicastAddress` and `Add/RemoveUnicastAddress()` (and multicast version), the fact that they add/remove the `NetifUnicastAddress` instance itself (and not copy it) is intentional and part of the design and public behavior of `Netif` class methods:
- We have two types of methods for adding/removing unicast addresses (similarly for multicast):
   -  `AddUnicastAddress()`/`RemoveUnicastAddress()`  (which are intended for core/internal use) and will directly use the passed-in `NetifUnicastAddress` instance itself.
   - `AddExternalUnicastAddress()`/`RemoveExternalUnicastAddress()` which will not use the passed-in `NetifUnicastAddress` instance and will create a copy of it.
- So if in other modules/classes, we rely on the fact that `Add/RemoveUnciastAddress()` is using the entry, it is consistent and part of the intended known public behavior of API/methods of `Netif` (though I would say, we have room for improvement in documenting this more clearly in the `Netif` module).

Reply:
Thanks for your clarification. What you said makes much sense.

If we are going to use address comparing rather than pointer comparing, we need to make sure that `DuaManager::GetDomainUnicastAddress()` always return `NULL` when DUA is not added to `Netif`, meaning we need to correctly set `mDomainUnicastAddress.mValid` at all cases, which I think is not true for now:

- When `DUA` is removed from `Netif` due to domain prefix change,  `mDomainUnicastAddress.mValid` is not set to `false`, thus  `DuaManager::GetDomainUnicastAddress()` can still return the `DUA` address which is not correct. 
https://github.com/openthread/openthread/blob/7b0cb984455731bbaf2eb51188baf5d225a2ee41/src/core/thread/dua_manager.cpp#L62-L66

- When `SLAAC` failed to generate a `DUA`, `mDomainUnicastAddress.mValid` is not set to `false` either. 
https://github.com/openthread/openthread/blob/7b0cb984455731bbaf2eb51188baf5d225a2ee41/src/core/thread/dua_manager.cpp#L80-L88

Reply:
Can we just compare `addr` with `&Get<DuaManager>().mDomainUnicastAddress` ?

Reply:
@abtink  @simonlingoogle Thanks for the discussion here! 
Agree that it would be better to check the address value directly for safe. 
To keep the consistence of core/internal unicast address usage, I pushed the second commit to resolve your concerns:
1) Simply return the reference of the address in `DuaManager().GetDomainUnicastAddress()`;
2) Use `Get<ThreadNetif>().IsUnicastAddress()` to determine whether it is a valid address successfully added to the interface.

Please take a look.

Reply:
Sounds great and simple. Thanks